### DISCUSSION TITLE
How to get the editor top level node only on demand ?

### DISCUSSION THREAD

**Discussion by Xuffasch** [Milkdown/milkdown#61] [Q&A] [answered]
Hello,

I'm interested in getting the current doc node in Milkdown on-demand, such as when I click on a button for example, but the listener is only sending the doc node on updates.

Is there any way I can do that ?

I'm currently creating a ProseMirror editor to visually apply structure to plain text for non-tech people. I would be very interested to translate that text structure to markdown for static site generation but also translate existing markdown text into my visual formatting with nodeviews. I was thinking that I would copy the doc node from Milkdown then use it in my ProseMirror editor. 

I don't want to have to update the document to trigger the retrieval of the latest version of the doc node.

I hope I was clear in my explanation.

**Comment by Saul-Mirone** [comment]
```typescript
const milkdown = new Editor().create();

const node = milkdown.getNode();
```
Do you mean something like this?

**Reply by Xuffasch** [reply]
Yes, something like this. 

This is currently not possible, isn't it ? I see that the Editor().create() returns undefined for now.

I thought that Milkdown editor would work like Prosemirror EditorView from which you can get the doc using editorView.state.doc

**Comment by Saul-Mirone** [comment, marked as answer]
In fact you can get the prosemirror view:

```typescript
const editor = new Editor()

editor.use(...).create()

const view = editor.view
```
---

The version 4.6 has been released and you can use following code:

> BTW, I still need to update doc about this

```typescript
import { Editor, editorViewOptionsCtx, editorViewCtx } from '@milkdown/core';

async function createEditor() {
    const newEditor = await new Editor()
        .config(ctx => {
            ctx.update(editorViewOptionsCtx, prev => ({ ...prev, root: root2, defaultValue: content2 }))
        })
        .use([directiveRemarkPlugin, titleNode])
        .use(commonmark)
        .create();

    const view = newEditor.action(ctx => ctx.get(editorViewCtx));

    // do whatever you want.
    const doc = view.state.doc;
}
```

**Reply by Xuffasch** [reply]
Hello, thanks for your answers.

I used the above method : setup the editor first then add the plugins and create the editor. (By the way, we need to at least provide the root and the defaultValue to make it work). I can see the editorView and reach the doc node.

```js
 let newEditor = new Editor({
      root: root2,
      defaultValue: content2,
    });
    newEditor
      // .use([directiveRemarkPlugin, new TitleNode()])
      .use(commonmark)
      .create();
    const view = newEditor.view;
    console.log("blank editor : ", newEditor);
    console.log("blank editor view : ", view);
    console.log("blank.editor editorView state doc : ", view.state.doc);
  });
```

However, if I want to insert plugins such as the one you showed us to implement custom syntax, i'm getting a blocking error:

state.ts:23 Uncaught Error
    at State._State_matchTarget2 (state.ts:23)
    at State._State_runNode2 (state.ts:29)
    at state.ts:84
    at Array.forEach (<anonymous>)
    at State.next (state.ts:84)
    at runner (paragraph.ts:22)
    at State._State_runNode2 (state.ts:32)
    at state.ts:84
    at Array.forEach (<anonymous>)
    at State.next (state.ts:84)

If I setup the editor this way, the plugin works: 
```js
    let milkeditor = new Editor({
      root,
      defaultValue: content,
    })
      .use([directiveRemarkPlugin, new TitleNode()])
      .use(commonmark)
      .create();


```

but the resulting editor is undefined so I cannot reach the prose mirror editorView and its state.

Please tell me if I'm clear. I've been working on Prose for a month and half and still it's difficult for me.

**Reply by Saul-Mirone** [reply]
There are some syntax issues in your sample code so I couldn't tell where the problem is. In fact there won't be any difference wether you `use` the plugins following the constructor or later. Maybe you can provide a sample project so I can clone it to see where the problem is. Or if you are not in a hurry, there will be an exciting major release version with much improvements in this week.

**Reply by Saul-Mirone** [reply]
You can have a taste of it at here: https://github.com/Saul-Mirone/milkdown/blob/functional-plugin/packages/preset-commonmark/app/main.ts

**Reply by Saul-Mirone** [reply]
Hi, @Xuffasch. The version 4.6 has been released and you can use following code:

> BTW, I still need to update doc about this

```typescript
import { Editor, editorViewOptionsCtx, editorViewCtx } from '@milkdown/core';

async function createEditor() {
    const newEditor = await new Editor()
        .config(ctx => {
            ctx.update(editorViewOptionsCtx, prev => ({ ...prev, root: root2, defaultValue: content2 }))
        })
        .use([directiveRemarkPlugin, titleNode])
        .use(commonmark)
        .create();

    const view = newEditor.action(ctx => ctx.get(editorViewCtx));

    // do whatever you want.
    const doc = view.state.doc;
}
```

**Reply by adaboese** [reply]
Thank you. This really works. The only challenge I have now is that `selector` does not update when `listener.markdownUpdated` is triggered. Just wondering if I introduced an error or if your provided example does not account for that?

**Reply by adaboese** [reply]
I was missing `ctx.get(editorViewCtx).state.doc`. Thank you!
